### ISSUE TITLE
Questions related to the anonymizer training

### ISSUE THREAD

**Issue by Sara-Esam** [UCF-CRCV/TeD-SPAD#4]
Dear Joseph, 

I have got a couple of questions which I wish you could help me answer .. 
### **Question (1): Related to the anonymizer training**
I was trying to reproduce the anonymization results by retraining the models using the provided settings as they are.

However, after training for 100 epochs, the accuracy of action model (fb) reached ~33% and the privacy model was not producing any useful visual results. 

When I visualize the maps saved every epoch, I see the following 
epoch 30 
![image](https://github.com/UCF-CRCV/TeD-SPAD/assets/78351047/df5429ea-123f-4471-a0a0-5f13abf9df64)

epoch 50 
![image](https://github.com/UCF-CRCV/TeD-SPAD/assets/78351047/dcf97fc7-5d22-40a1-b103-8b547781f70c)

epoch 80 
![image](https://github.com/UCF-CRCV/TeD-SPAD/assets/78351047/f035dd92-2403-4fb3-9834-8d404b16bc8f)

epoch 100 
![image](https://github.com/UCF-CRCV/TeD-SPAD/assets/78351047/db186330-6ef3-4262-86f3-26a8d097feff)

I did not use any saved model files for loading the ft and fb, since these are missing on the given drive link. But I saw that pre-trained model versions are already used. 

These are the parameters I used for the anonymizer
```
import numpy as np
import math
import os.path


# Job parameters.
run_id = 'baseline_anonymization'
arch_ft = 'largei3d'
arch_fa = 'unet++'
arch_fb = 'r50'
saved_model_fa = None # os.path.join('..', 'saved_models', 'fa_recon.pth')
saved_model_ft = None
saved_model_fb = None #os.path.join('..', 'saved_models', 'fb_ssl.pth')

# Dataset parameters.
num_classes = 102
num_frames = 16
fix_skip = 2
num_modes = 2
num_skips = 1
data_percentage = 0.1

# Number of VISPR privacy attributes.
num_pa = 7
data_percentage_vispr = 1.0

# Training parameters.
batch_size = 4
batch_size_vispr = 4
v_batch_size = 4
num_workers = 120
learning_rate = 1e-4
num_epochs = 100
loss = 'ce'
temporal_loss = 'trip'
warmup_array = list(np.linspace(0.01, 1, 10) + 1e-9)
warmup = len(warmup_array)
momentum = 0.9
weight_decay = 1e-4
lr_warmup_method = 'linear'
lr_warmup_epochs = 10
lr_warmup_decay = 0.001
lr_gamma = 0.1
lr_milestones = [20, 30, 40]
lr_patience = 2
lr_reduce_factor = 2
lr_scheduler = 'cosine' # 'patience_based' #'loss_based' (default) #'cosine'
cosine_lr_array = list(np.linspace(0.01,1, 5)) + [(math.cos(x) + 1)/2 for x in np.linspace(0,math.pi/0.99, num_epochs-5)]
val_freq = 3
opt_type = 'adam' # 'sgd' # 'adamw'
ft_dropout = 0

# Anonymization training parameters.
# Scaled lr per model.
learning_rate_fa = 0.4*learning_rate
learning_rate_fb = 1.0*learning_rate
learning_rate_ft = 1.0*learning_rate
ft_loss_weight = 0.7
fb_loss_weight = 1.0
temporal_loss_weight = 0.1
weight_inv = 0.0
triplet_loss_margin = 1
temporal_distance = 4

# Validation augmentation params.
hflip = [0]
cropping_facs = [0.8]
weak_aug = False
no_ar_distortion = False
aspect_ratio_aug = False

# Training augmentation params.
reso_h = 224
reso_w = 224
min_crop_factor_training = 0.6
temporal_align = False

# Tracking params.
wandb = False
 
```
Based on your deep knowledge of the code, what would be the suspect for this issue? 

### **Question (2): Related to fb model choice**
The loading of the model fb seems a bit ambiguous for me, I was unable to tell the difference between loading the model with `ssl=True` or  `ssl=False` 
I also notice that with `ssl=True`, the ResNet model used was initialized randomly whereas the second case uses some pre-trained version 

```
# Load pretrained VISPR privacy attribute model.
def load_privacy_ssl():
    # MLP layer.
    class MLP(nn.Module):
        def __init__(self, final_embedding_size=128, use_normalization=True):
            super(MLP, self).__init__()
            self.final_embedding_size = final_embedding_size
            self.use_normalization = use_normalization
            self.fc1 = nn.Linear(2048, 2048, bias=True)
            self.relu = nn.ReLU(inplace=True)
            self.fc2 = nn.Linear(2048, self.final_embedding_size, bias=True)

        def forward(self, x):
            x = self.relu(self.fc1(x))
            x = nn.functional.normalize(self.fc2(x), p=2, dim=1)
            return x

    # Can adapt to different architectures.
    resnet_model = resnet50(weights=None)
    resnet_model.fc = nn.Identity()
    mlp = MLP()
    #################
    # Do I need this?
    # state_dict = mlp.state_dict()
    # mlp.load_state_dict(state_dict, strict=True)
    #################
    fb_model = nn.Sequential(resnet_model, mlp)
    # saved_dict = torch.load(saved_model_file)
    # fb_model.load_state_dict(saved_dict['fb_model_state_dict'], strict=True)
    # print(f'fb_model loaded from {saved_model_file} successfully!')
    return fb_model


# Build ResNet model for privacy prediction.
def build_resnet_predictor(num_classes=7, pretrained=True):
    if pretrained:
        model = resnet50(weights=ResNet50_Weights.DEFAULT)
    else:
        model = resnet50(weights=None)

    # Replace fc layer to get desired output.
    in_features = model.fc.in_features
    model.fc = nn.Linear(in_features, num_classes, bias=True)

    return model 
```

Which of these should be used if I am training for the anonymization? 

### **Question (3): What is meant by the number of modes specified in for validation ?** 

Looking forward to your response with many Thanks in advance!

**Comment by joefioresi718** [comment]
Thank you for your very detailed question! I apologize for such a late response, I did not get a notification for this.

1. The training is an adversarial minimax optimization, so it does not converge properly. Instead, a model is chosen at some point during training where performance is roughly maintained and anonymization quality is apparent. This is not entirely straightforward and can drastically change from run to run. Also, the method is very sensitive to hyperparameters. My mistake, 
I believe I had used a batch size of 8 instead of the 4 I have in the file. Using that, or with some adjustments to the learning rate and a few tries, you should be able to get solid results. I have uploaded the model weights to the OneDrive link.
2. The SSL version is used during training and outputs a projected embedding instead of an attribute prediction, as we do not allow access to attribute labels. In practice, we found that the initialization here was not that important, but I have provided the weights I had used (fb_ssl.pth) in the Model Weights folder. The second, non-SSL variant is used solely for the 'privacy_training', which is where we evaluate the quality of the anonymizer model by training and evaluating an attribute predictor (non-SSL ResNet).
3. The number of modes is simply the amount of clips used for validation (standard practice to use more than 1 in action recognition).

Let me know if you have any more questions, thank you for your interest in my work.
